### PR TITLE
DLC Refund Transaction Test

### PR DESCRIPTION
**Pull request by nkohen**
This PR makes the `timeout` type a `BlockStamp` instead of an `Int`.

It also adds an execution case with new tests for spending the refund transaction.

### CODE REVIEW

File: core/src/main/scala/org/bitcoins/core/protocol/BlockStamp.scala [hunk 1]
```diff
@@ -14,22 +14,43 @@ sealed trait BlockStamp {
...
+    override def toUInt32: UInt32 = time
+  }
+
+  /** This is Tue Nov  5 00:53:20 1985 UTC
+    * @see [[https://github.com/bitcoin/bitcoin/blob/master/src/script/script.h#L39]]
+    */
+  final val LOCKTIME_THRESHOLD: Int = 500000000
+
+  /** @see [[https://github.com/bitcoin/bips/blob/master/bip-0065.mediawiki#detailed-specification]] */
+  def apply(timelockNumber: Int): BlockStampWithFuture = {
```
Comment (benthecarman) [inline core/src/main/scala/org/bitcoins/core/protocol/BlockStamp.scala:52]: Should `timelockNumber` be a`UInt32`

File: dlc/src/main/scala/org/bitcoins/dlc/BinaryOutcomeDLCWithSelf.scala [hunk 2]
```diff
@@ -370,97 +373,253 @@ case class BinaryOutcomeDLCWithSelf(
...
-              finalLocalPrivKey
+  /** Constructs and executes on the unilateral spending branch of a DLC
+    *
+    * @return Each transaction published and its spending info
+    */
+  def executeUnilateralDLC(
+      oracleSigF: Future[SchnorrDigitalSignature],
+      local: Boolean,
+      messengerOpt: Option[BitcoinP2PMessenger] = None): Future[DLCOutcome] = {
+    setupDLC(messengerOpt).flatMap {
```
Comment (Christewart) [inline dlc/src/main/scala/org/bitcoins/dlc/BinaryOutcomeDLCWithSelf.scala]: setup shouldn't be happening in the `executeUnilateralDLC()` method. You are co-mingling test code and source code.

File: dlc/src/main/scala/org/bitcoins/dlc/BinaryOutcomeDLCWithSelf.scala [hunk 3]
```diff
@@ -370,97 +373,253 @@ case class BinaryOutcomeDLCWithSelf(
...
+          // The prefix other refers to remote if local == true and local otherwise
+          val output = cet.outputs.head
+          val otherOutput = cet.outputs.last
+
+          // Spend the true case on the correct CET
+          val cetSpendingInfo = ConditionalSpendingInfo(
+            TransactionOutPoint(cet.txIdBE, UInt32.zero),
+            output.value,
+            output.scriptPubKey.asInstanceOf[ConditionalScriptPubKey],
+            Vector(cetPrivKey, ECPrivateKey(oracleSig.s)),
```
Comment (Christewart) [inline dlc/src/main/scala/org/bitcoins/dlc/BinaryOutcomeDLCWithSelf.scala]: this seems kind of hacky

File: dlc/src/main/scala/org/bitcoins/dlc/BinaryOutcomeDLCWithSelf.scala [hunk 4]
```diff
@@ -231,7 +232,7 @@ case class BinaryOutcomeDLCWithSelf(
       requiredSigs = 2,
       pubKeys = Vector(cetRemotePrivKey.publicKey, sigPubKey))
     val timeoutSPK = CLTVScriptPubKey(
-      locktime = ScriptNumber(timeout),
+      locktime = ScriptNumber(timeout.toUInt32.toInt),
```
Comment (Christewart) [inline dlc/src/main/scala/org/bitcoins/dlc/BinaryOutcomeDLCWithSelf.scala]: I would add a `toScriptNumber()` method on `BlockStamp`

File: dlc/src/main/scala/org/bitcoins/dlc/BinaryOutcomeDLCWithSelf.scala [hunk 5]
```diff
@@ -323,15 +324,17 @@ case class BinaryOutcomeDLCWithSelf(
...
+      fundingTx: Transaction,
+      fundingSpendingInfo: MultiSignatureSpendingInfo,
+      cetWinLocal: Transaction,
+      cetLoseLocal: Transaction,
+      cetWinRemote: Transaction,
+      cetLoseRemote: Transaction,
+      refundTx: Transaction
+  )
+
+  def setupDLC(messengerOpt: Option[BitcoinP2PMessenger]): Future[SetupDLC] = {
```
Comment (Christewart) [inline dlc/src/main/scala/org/bitcoins/dlc/BinaryOutcomeDLCWithSelf.scala]: Should this be test code?

File: dlc/src/main/scala/org/bitcoins/dlc/BinaryOutcomeDLCWithSelf.scala [hunk 6]
```diff
@@ -370,97 +373,253 @@ case class BinaryOutcomeDLCWithSelf(
...
+    }
+  }
+
+  /** Constructs and executes on the refund spending branch of a DLC
+    *
+    * @return Each transaction published and its spending info
+    */
+  def executeRefundDLC(
+      messengerOpt: Option[BitcoinP2PMessenger] = None): Future[DLCOutcome] = {
+    setupDLC(messengerOpt).flatMap {
```
Comment (Christewart) [inline dlc/src/main/scala/org/bitcoins/dlc/BinaryOutcomeDLCWithSelf.scala]: setup should be in test code

**Review by Christewart — COMMENTED** [review]
Two things

1. It's pretty clear we are co-mingling test code and source code at this point, all setup stuff should be moved into test cases. This code is going to have to be reworked pretty significantly in the future as you don't have access to the other persons private keys. If this stuff was setup properly in tests, you can have the test case setup the remote side of the DLC and the local side of the DLC, using the same code. Right now you just assume everything is available in src.

2. Can you please use named parameters (`alt+enter` in intellij) so that it is easier for reviewers to review your code?

File: dlc/src/main/scala/org/bitcoins/dlc/BinaryOutcomeDLCWithSelf.scala (hunk 3 above)
Comment (nkohen) [reply dlc/src/main/scala/org/bitcoins/dlc/BinaryOutcomeDLCWithSelf.scala]: Yup, DLCs are essentially just this hack (assuming you are referring to `ECPrivateKey(oracleSig.s)`

**Comment by nkohen** [comment]
I plan on removing `BitcoinP2PMessenger` in the subsequent PR (this should de-futurify things).

And of course, separating this out into two nodes with different information is work for a future PR once construction is finalized.

File: dlc/src/main/scala/org/bitcoins/dlc/BinaryOutcomeDLCWithSelf.scala (hunk 3 above)
Comment (Christewart) [reply dlc/src/main/scala/org/bitcoins/dlc/BinaryOutcomeDLCWithSelf.scala]: I would like to see if we can do something with the `Sign` interface, but i guess this isn't the time for that.

**Pull request merged**